text/0000-fabric-private-chaincode.md

```diff
@@ -0,0 +1,472 @@
+
+* Step 4: Ordering
+
+	![Ordering](../images/fpc/high-level/FPC-Ordering.png)
+
+	With FPC we follow the normal Ordering service. While the Orderer is unmodified in FPC, the ordered transactions broadcast to the Peers in the Channel for validation now include attested endorsements. Note that we encourage the use of a BFT-based ordering service as FPC requires the Ordering service to be trusted.
+
+* Step 5: Validation
+
+	![Validate](../images/fpc/high-level/FPC-Validate.png)
+
+	As the Peers in the Channel receive the block of transactions, they perform the standard Fabric validation process.
+	In addition, a custom validation plugin in the Peer is responsible to verify FPC transactions. In particular, the FPC Validator queries its local FPC Registry to retrieve the enclave signature verification key and check that the signature was produced in an actual and correct Chaincode Enclave. This query retrieves the stored Attestation report associated with the Public Key of the Chaincode Enclave that produced and signed the transaction. The Attestation report is checked to verify the details of the Chaincode Enclave, affirming the validity of the Enclave.
+
+* Step 6: Trusted Ledger Revalidation
+
+	![Revalidate](../images/fpc/high-level/FPC-Revalidate.png)
+
+	In addition to the standard validation and commit process, the Peer also forwards all committed blocks to the Ledger Enclave in order to establish a full and current trusted view of the ledger. 
+	The same validation steps described above are repeated inside the Enclave, and then the transaction is committed to the trusted version of the Ledger inside this Enclave. This constitutes an update to the World State Integrity Metadata: for each Key Value Pair of World State, a second Key Value Pair is stored in the Trusted Ledger containing Integrity Metadata (a cryptographic hash value) along with channel-specific details needed to verify that the transaction was produced by a valid authorized participant in the channel.
+
+* Step 7: Commitment
+
+	![Commit](../images/fpc/high-level/FPC-Commit.png)
+
+	If these validation processes succeed, the FPC transaction is marked valid.
+	The Peer continues with committing the transactions to the local ledger and applies all valid transactions (i.e. the write set) to the World State. This completes the transaction.
+
+
+## Subsequent Query and Validation of World State Data by FPC Chaincodes
+
+Subsequent to the transaction described above: When any FPC Chaincode accesses the World State, the FPC Shim is responsible not only for passing the query to the Peer but also querying the Ledger Enclave for the Integrity Metadata for the Key in question. It performs a cryptographic hash of the Key Value Pair of World State returned by the Peer and checks it against the Integrity Metadata from the Ledger Enclave for that key. If the hashes do not match, it detects an integrity violation.
```

Comment:
> It performs a cryptographic hash of the Key Value Pair of World State returned by the Peer and checks it against the Integrity Metadata from the Ledger Enclave for that key. If the hashes do not match, it detects an integrity violation.

There is likely a better way of doing it, such as consistently maintaining an accumulator over the world state, and keeping track only of its value in the FPC persistent storage and when you query the peer, the peer provides a membership proof for the accumulator, or a non membership proof of the accumulator. 

One construction for an accumulator is a sorted Merkle tree.

Reply:
For keys which do not exist, there will be no metadata which would wrongly validate its existance; for keys which do exist but a malicious peer tries to hide; we cross-check also that non-existing tags indeed do not exist as part of our meta-data.

Regarding merkle-tree, this was implicitly referred to as part of the "Trusted Ledger State Snapshot:" in the road-map features section.

Reply:
> For keys which do not exist, there will be no metadata which would wrongly validate its existance

But don't you need to prove that it is indeed the case? Otherwise I can simply trick the FPC to compute the wrong output, no?

Reply:
what i forgot to mention: when the trusted_ledger does not find meta-data it will notify the chaincode enclave of that fact in a crypto-graphically protected way (over the secure channel between trusted ledger enclave and the chaincode enclave). That way the chaincode enclave can validate responses from the untrusted peer, whether it provides a value or claims no value for a given key exists ..